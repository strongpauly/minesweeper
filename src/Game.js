import React, { Component } from 'react';
import Cell from './Cell';
import Set from 'betterset';
import PropTypes from 'prop-types';

class Game extends Component {

  static propTypes = {
    numMines: PropTypes.number,
    width: PropTypes.number,
    height: PropTypes.number,
    onRestart: PropTypes.func
  }

  constructor(props) {
    super(props);
    let mines = new Map();
    let adjacentCount = new Map();
    for (let mineNumber = 0; mineNumber < props.numMines; mineNumber ++) {
      let setMine = false;
      while(!setMine) {
        let x = Math.round(Math.random() * (props.width - 1));
        let y = Math.round(Math.random() * (props.height - 1));
        let key = this.getCellKey(x, y);
        if(!mines.has(key)) {
          mines.set(key, {x: x, y:y});
          setMine = true;
        }
      }
    }
    mines.forEach(mine => {
      for (let x = mine.x - 1; x <= mine.x + 1; x++) {
        for (let y = mine.y - 1; y <= mine.y + 1; y++) {
          if( (x !== mine.x || y !== mine.y) && this.validCoord(x, y)) {
            let key = this.getCellKey(x, y);
            let count = adjacentCount.get(key);
            if (count === undefined) {
              count = 0;
            }
            count ++;
            adjacentCount.set(key, count);
          }
        }
      }
    });
    this.state = {
      exploded: false,
      checked: new Set(),
      marked: new Set(),
      adjacentCount: adjacentCount,
      mines: mines,
      completed: false
    };

    this.onCheck = this.onCheck.bind(this);
    this.onMark = this.onMark.bind(this);
    this.restart = this.restart.bind(this);
  }

  /**
   * Helper function to determine whether a coordinate appears inside the bounds of the grid.
   */
  validCoord(x, y) {
    return x >= 0 && y >= 0 && x < this.props.width && y < this.props.height;
  }

  getCellKey(x, y) {
    return x + ',' + y;
  }

  getCellCoord(key) {
    let split = key.split(',');
    return {x:parseInt(split[0], 10), y: parseInt(split[1], 10)};
  }

  startTimer() {
    if(!this.timerId) {
      this.setState({time: 1});
      this.timerId = setInterval(() => {
        this.setState((previousState) => ({
          time:previousState.time + 1
        }));
      }, 1000);
    }
  }

  stopTimer() {
    if(this.timerId) {
      clearInterval(this.timerId);
    }
  }

  componentWillUnmount() {
    this.stopTimer();
  }

  onCheck(cellX, cellY) {
    if(!this.state.completed) {
      this.startTimer();
      let key = this.getCellKey(cellX, cellY);
      if(this.state.mines.has(key)) {
        this.setState({exploded:true, completed:true});
        this.stopTimer();
      } else {
        this.doCheck(key);
        if(this.hasWon()) {
          this.setState({completed:true});
          this.stopTimer();
        }
        this.setState({
          checked: this.state.checked,
          marked: this.state.marked
        });
      }
    }
  }

  onMark(cellX, cellY, mark) {
    if(!this.state.completed) {
      this.startTimer();
      let key = this.getCellKey(cellX, cellY);
      if(mark) {
        this.state.marked.add(key);
      } else {
        this.state.marked.delete(key);
      }
      this.setState({marked: this.state.marked});
    }
  }

  doCheck(cellKey) {
    let checking = new Set([cellKey]);
    while (checking.size > 0) {
      let key = checking[Symbol.iterator]().next().value;
      checking.delete(key);
      //May have already been checked in this iteration.
      if(!this.state.checked.has(key)) {
        this.state.checked.add(key);
        //Stop when we get close to a mine.
        if(!this.state.adjacentCount.has(key)) {
          let coord = this.getCellCoord(key);
          checking.addAll(
              //       x-1 x   x+1
              // y-1   X   X   X
              // y     X   O   X
              // y+1   X   X   X
              [
                  {x: coord.x - 1, y: coord.y - 1},
                  {x: coord.x - 1, y: coord.y},
                  {x: coord.x - 1, y: coord.y + 1},
                  {x: coord.x, y: coord.y - 1},
                  {x: coord.x, y: coord.y + 1},
                  {x: coord.x + 1, y: coord.y - 1},
                  {x: coord.x + 1, y: coord.y},
                  {x: coord.x + 1, y: coord.y + 1}
              ].filter(newCoord => {
                let newKey = this.getCellKey(newCoord.x, newCoord.y);
                return this.validCoord(newCoord.x, newCoord.y) &&
                  !this.state.mines.has(newKey) &&
                  !this.state.checked.has(newKey) &&
                  !this.state.marked.has(newKey);
              }).map(newCoord => this.getCellKey(newCoord.x, newCoord.y))
          );
        }
      }
    }
  }

  restart() {
    this.props.onRestart();
  }

  hasWon() {
      //Only mines showing.
    return (this.props.width * this.props.height) - this.state.checked.size === this.props.numMines;
  }

  render() {
    let widthArray = new Array(this.props.width).fill();
    let heightArray = new Array(this.props.height).fill();
    let cells = heightArray.map((emptyY, y) => {
      let row = widthArray.map((emptyX, x) => {
        let key = this.getCellKey(x, y);
        return <Cell key={key} x={x} y={y}
            mine={this.state.exploded && this.state.mines.has(key)}
            checked={this.state.checked.has(key)}
            marked={this.state.marked.has(key)}
            adjacentCount={this.state.adjacentCount.get(key)}
            onCheck={this.onCheck}
            onMark={this.onMark}></Cell>;
      });
      return <tr className="row" key={'row' + y}>{row}</tr>;
    });
    return <div className="gridContainer">
            <div>
                <div className="header">
                    <div className="numMines">{this.props.numMines - this.state.marked.size}</div>
                    <div className="status" onClick={this.restart}>{this.state.exploded ? String.fromCharCode(9760) : this.state.completed ? 'âœ”' : String.fromCharCode(9822)}</div>
                    <div className="timer">{this.state.time || ' '}</div>
                </div>
                <table className={this.state.completed ? 'grid completed' : 'grid'}>
                    <tbody>{cells}</tbody>
                </table>
            </div>
        </div>;
  }
}

export default Game;
