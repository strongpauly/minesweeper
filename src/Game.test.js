import React from 'react';
import Game from './Game';
import {shallow} from 'enzyme';

/* eslint-env jest */

describe('<Game>', () => {

  it('renders without crashing', () => {
    shallow(<Game width={10} height={10} numMines={10}/>);
  });

  it('has header', () => {
    const cell = shallow(<Game width={0} height={0} numMines={0}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('.header')).toHaveLength(1);
  });

  it('creates grid based on width and height', () => {
    const cell = shallow(<Game width={2} height={2} numMines={2}/>);
    expect(cell).toMatchSnapshot();
    expect(cell.find('tr')).toHaveLength(2); //Height 2.
    expect(cell.find('Cell')).toHaveLength(4); //Height 2 X Width 2.
  });

  it('shows number of mines in header', () => {
    const cell = shallow(<Game width={1} height={2} numMines={1}/>);
    expect(cell).toMatchSnapshot();
    const numMines = cell.find('.header .numMines');
    expect(numMines).toHaveLength(1);
    expect(numMines.text()).toEqual('1');
  });
});
