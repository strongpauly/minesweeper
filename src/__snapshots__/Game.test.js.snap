// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game> creates grid based on width and height 1`] = `
<div
  className="gridContainer"
>
  <div>
    <div
      className="header"
    >
      <div
        className="numMines"
      >
        2
      </div>
      <div
        className="status"
        onClick={[Function]}
      >
        ♞
      </div>
      <div
        className="timer"
      >
         
      </div>
    </div>
    <table
      className="grid"
    >
      <tbody>
        <tr
          className="row"
        >
          <Cell
            adjacentCount={1}
            checked={false}
            marked={false}
            mine={false}
            onCheck={[Function]}
            onMark={[Function]}
            x={0}
            y={0}
          />
          <Cell
            adjacentCount={2}
            checked={false}
            marked={false}
            mine={false}
            onCheck={[Function]}
            onMark={[Function]}
            x={1}
            y={0}
          />
        </tr>
        <tr
          className="row"
        >
          <Cell
            adjacentCount={2}
            checked={false}
            marked={false}
            mine={false}
            onCheck={[Function]}
            onMark={[Function]}
            x={0}
            y={1}
          />
          <Cell
            adjacentCount={1}
            checked={false}
            marked={false}
            mine={false}
            onCheck={[Function]}
            onMark={[Function]}
            x={1}
            y={1}
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Game> has header 1`] = `
<div
  className="gridContainer"
>
  <div>
    <div
      className="header"
    >
      <div
        className="numMines"
      >
        0
      </div>
      <div
        className="status"
        onClick={[Function]}
      >
        ♞
      </div>
      <div
        className="timer"
      >
         
      </div>
    </div>
    <table
      className="grid"
    >
      <tbody />
    </table>
  </div>
</div>
`;

exports[`<Game> shows number of mines in header 1`] = `
<div
  className="gridContainer"
>
  <div>
    <div
      className="header"
    >
      <div
        className="numMines"
      >
        1
      </div>
      <div
        className="status"
        onClick={[Function]}
      >
        ♞
      </div>
      <div
        className="timer"
      >
         
      </div>
    </div>
    <table
      className="grid"
    >
      <tbody>
        <tr
          className="row"
        >
          <Cell
            checked={false}
            marked={false}
            mine={false}
            onCheck={[Function]}
            onMark={[Function]}
            x={0}
            y={0}
          />
        </tr>
        <tr
          className="row"
        >
          <Cell
            adjacentCount={1}
            checked={false}
            marked={false}
            mine={false}
            onCheck={[Function]}
            onMark={[Function]}
            x={0}
            y={1}
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
